src/sub.rs

```diff
@@ -31,6 +116,7 @@ impl Drop for SubSocket {
 
 impl SubSocket {
     pub async fn subscribe(&mut self, subscription: &str) -> ZmqResult<()> {
```

Comment:
I guess subscribe/unsubscribe impl can be moved to backend to avoid copy/pasting code around

Reply:
Could you elaborate a little bit on that.

Reply:
I don't really like that we have 2 places where we create a subscribe message. So I would look for a way to move this code to some common function.

Reply:
I went with the following approach but cannot borrow due to this: `E0596 cannot borrow data in an 'Arc' as mutable`
```rust
pub enum SubBackendMessageType {
    UNSUBSCRIBE = 0,
    SUBSCRIBE = 1
}

impl SubSocket {
    pub async fn subscribe(&mut self, subscription: &str) -> ZmqResult<()> {
	self.backend.handle_subscription(
	    subscription, SubBackendMessageType::SUBSCRIBE).await
    }

    pub async fn unsubscribe(&mut self, subscription: &str) -> ZmqResult<()> {
	self.backend.handle_subscription(
	    subscription, SubBackendMessageType::UNSUBSCRIBE).await
    }
}
```
and defined `handle_subscription` on the backend
```rust
    pub async fn handle_subscription(
	&mut self, subscription: &str, msg_type: SubBackendMessageType) -> ZmqResult<()> {
        let mut buf = BytesMut::with_capacity(subscription.len() + 1);
        buf.put_u8(msg_type as u8);
        buf.extend_from_slice(subscription.as_bytes());

        let message: ZmqMessage = ZmqMessage::from(buf.freeze());
        for mut peer in self.peers.iter_mut() {
            peer.send_queue
                .send(Message::Message(message.clone()))
                .await?;
        }
        self.subs.lock().insert(subscription.to_string());

	Ok(())
	
    }
```
How would you define a common function in the backend and call it from the `SubSocket`

Thanks

Reply:
1) Backend is shared between threads so it cannot have methods that take `&mut self`.
2) I would simple create an assosiated function that would create a subscribe message in it.
```
fn subscribe(subscription: &str) -> ZmqMessage {
        let mut buf = BytesMut::with_capacity(subscription.len() + 1);
        buf.put_u8(msg_type as u8);
        buf.extend_from_slice(subscription.as_bytes());

        buf.freeze().into()
}
```
And use it where we need to create a message.